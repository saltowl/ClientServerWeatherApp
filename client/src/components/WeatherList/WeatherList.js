import React from 'react';
import Weather from '../Weather/Weather';

class WeatherList extends React.Component {
  constructor(props) {
    super(props);
  }
  
  render() {
    if (Array.isArray(this.props.cities)) {
      return (
        <div className="WeatherList">
            {this.props.cities.map(city => (
                <Weather key={city.id} data={city} fetchWeather={() => this.props.fetchWeatherByName(city.id, city.name)} deleteCity={() => this.props.deleteCity(city.id)} />
            ))}
        </div>
      );
    } else {
      return (
        <div className="WeatherList"></div>
      );
    }
  }
}

export default WeatherList;
