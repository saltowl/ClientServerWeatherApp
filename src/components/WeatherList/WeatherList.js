import React from 'react';
import Weather from '../Weather/Weather';

class WeatherList extends React.Component {
  constructor(props) {
    super(props);
  }
  
  render() {
    return (
      <div className="WeatherList">
          {this.props.cities.map(city => (
              <Weather key={city.id} data={city} fetchWeather={() => this.props.fetchWeather(city.id, city.name)} />
          ))}
      </div>
    );
  }
}

export default WeatherList;
