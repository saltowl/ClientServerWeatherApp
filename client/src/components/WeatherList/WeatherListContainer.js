import { connect } from "react-redux";
import * as actions from "../../store/actions";
import axios from "axios";
import WeatherList from "./WeatherList";
import { getCities } from "../../store/reducers";
import { API } from "../../constants";

const fetchWeatherByNameAction = (id, cityName) => {
  return dispatch => {
    dispatch(actions.fetchWeatherPending(id));
    axios
      .get(`${API}weather?city=${cityName}`)
      .then(response => {
        dispatch(actions.fetchWeatherSuccess(response.data, id));
      })
      .catch(error => {
        dispatch(actions.fetchWeatherError(error, id));
      });
  };
};

const deleteCityAction = id => {
  return dispatch => {
    dispatch(actions.deleteCityPending(id));
    axios
      .delete(`${API}favourites/${id}`)
      .then(response => {
        dispatch(actions.deleteCitySuccess(id));
      })
      .catch(error => {
        dispatch(actions.deleteCityError(error));
      });
  };
};

const fetchCitiesAction = () => {
  return dispatch => {
    dispatch(actions.fetchCitiesPending());
    axios
      .get(`${API}favourites`)
      .then(response => {
        const cities = response.data.map(city => ({ name: city.name, id: city._id }));
        dispatch(actions.fetchCitiesSuccess(cities));
      })
      .catch(error => {
        dispatch(actions.fetchCitiesError(error));
      });
  };
};

const mapStateToProps = state => ({
  cities: getCities(state)
});

const mapDispatchToProps = dispatch => ({
  fetchCities: () => dispatch(fetchCitiesAction()),
  fetchWeatherByName: (id, name) =>
    dispatch(fetchWeatherByNameAction(id, name)),
  deleteCity: id => dispatch(deleteCityAction(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(WeatherList);
