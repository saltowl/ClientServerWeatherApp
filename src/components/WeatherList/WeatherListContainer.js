import { connect } from 'react-redux';
import { fetchWeatherPending, fetchWeatherError, fetchWeatherSuccess, deleteCity } from '../../store/actions';
import axios from 'axios';
import WeatherList from './WeatherList';
import { getCities, getNextCityId } from '../../store/reducers';

export const fetchWeatherByNameAction = (id, cityName) => {
    return dispatch => {
        dispatch(fetchWeatherPending(id));
        axios
            .get(`https://api.openweathermap.org/data/2.5/weather?q=${cityName}&units=metric&APPID=96c2fc4713551153e7966978b449861a`)
            .then(response => {
                dispatch(fetchWeatherSuccess(response.data, id));
                console.log(response.data);
            })
            .catch(error => {
                dispatch(fetchWeatherError(error, id));
                console.log(error);
            });
    }
}


const mapStateToProps = (state) => ({
    cities: getCities(state),
    nextCityId: getNextCityId(state)
});

const mapDispatchToProps = (dispatch) => ({
    fetchWeatherByName: (id, name) => dispatch(fetchWeatherByNameAction(id, name)),
    deleteCity: id => dispatch(deleteCity(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(WeatherList);