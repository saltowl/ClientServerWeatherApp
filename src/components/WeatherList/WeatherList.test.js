import React from "react";
import WeatherList from "./WeatherList";
import renderer from "react-test-renderer";

const correctData = {
  base: "stations",
  clouds: { all: 75 },
  cod: 200,
  coord: { lon: 37.62, lat: 55.75 },
  dt: 1572515073,
  id: 524901,
  main: {
    temp: -2.4,
    pressure: 1021,
    humidity: 79,
    temp_min: -3,
    temp_max: -1.11
  },
  name: "Moscow",
  sys: {
    type: 1,
    id: 9027,
    country: "RU",
    sunrise: 1572496341,
    sunset: 1572530028
  },
  timezone: 10800,
  visibility: 10000,
  weather: [
    { id: 620, main: "Snow", description: "light shower snow", icon: "13d" }
  ],
  wind: { speed: 6, deg: 290 }
};

it("Filled WeatherList display correctly", () => {
  const cities = [
    { id: 1, pending: false, weather: correctData },
    { id: 2, pending: false, error: {} }
  ];
  const component = renderer.create(
    <WeatherList
      cities={cities}
      fetchWeatherByName={() => {}}
      deleteCity={() => {}}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("Empty WeatherList display correctly", () => {
  const component = renderer.create(
    <WeatherList
      cities={[]}
      fetchWeatherByName={() => {}}
      deleteCity={() => {}}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
