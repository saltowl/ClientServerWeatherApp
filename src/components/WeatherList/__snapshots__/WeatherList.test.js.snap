// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cities is a string, not an array 1`] = `
<div
  className="WeatherList"
/>
`;

exports[`Empty WeatherList display correctly 1`] = `
<div
  className="WeatherList"
/>
`;

exports[`Filled WeatherList display correctly 1`] = `
<div
  className="WeatherList"
>
  <div
    className="Weather"
    style={
      Object {
        "flexDirection": "column",
      }
    }
  >
    <div
      className="mainInfo"
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <div
        className="name"
      >
        Moscow
      </div>
      <div
        className=" cityMain"
      >
        <div
          className=""
        >
          -2.4°C
        </div>
        <img
          src="https://openweathermap.org/img/wn/13d.png"
          title="Light shower snow"
        />
      </div>
      <div
        className=""
      >
        <button
          className="circle"
          onClick={[Function]}
        >
          x
        </button>
      </div>
    </div>
    <div
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        className="row"
      >
        <div
          className="rowKey"
        >
          Wind:
        </div>
        <div
          className="rowVal"
        >
          <div
            id="wind"
          >
            <div
              id="N"
            >
              N
            </div>
            <div
              id="W"
            >
              W
            </div>
            <div
              id="windDirection"
            >
              <div
                style={
                  Object {
                    "OTransform": "rotate(290deg)",
                    "WebkitTransform": "rotate(290deg)",
                    "msTransform": "rotate(290deg)",
                    "transform": "rotate(290deg)",
                  }
                }
              >
                ↑
              </div>
            </div>
            <div
              id="E"
            >
              E
            </div>
            <div
              id="S"
            >
              S
            </div>
          </div>
          6 m/s
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="rowKey"
        >
          Clouds:
        </div>
        <div
          className="rowVal"
        >
          75
          %
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="rowKey"
        >
          Pressure:
        </div>
        <div
          className="rowVal"
        >
          766 mmHg
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="rowKey"
        >
          Humidity:
        </div>
        <div
          className="rowVal"
        >
          79%
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="rowKey"
        >
          Coords:
        </div>
        <div
          className="rowVal"
        >
          [
          37.62
          , 
          55.75
          ]
        </div>
      </div>
    </div>
  </div>
  <div
    className="Weather error"
  >
    <div>
      Poor connection. Please, try again later
    </div>
    <button
      className="circle"
      onClick={[Function]}
    >
      x
    </button>
  </div>
</div>
`;
